GitUI/BranchTreePanel/RepoObjectsTree.Nodes.cs

```diff
@@ -165,51 +174,45 @@ protected void ReloadNodes(Func<CancellationToken, Task> loadNodesTask)
 
                 ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
                 {
+                    Nodes.Clear();
```

Comment:
`ReloadNodes` should be an async function. `ThreadHelper.JoinableTaskFactory.RunAsync` should be used only at the beginning of the callstack and then all functions should be async functions if needed. This is how I understand the idea of using JFT.

Reply:
Nodes are cleared by each root node reload action, which causes unpleasant visual effect. I miss ReactJS + JS being single threaded - the code is clearer and easier to write.

Reply:
Hi @jbialobr, I've read over all your comments, and I think you may not have fully understood the change I made in https://github.com/gitextensions/gitextensions/commit/c9e47e4b3e79792b1afcc762e0164445c82606a6. I'll explain it here, then respond to your points.

Before https://github.com/gitextensions/gitextensions/commit/c9e47e4b3e79792b1afcc762e0164445c82606a6, There was a ReloadAsync function on RepoObjectsTree, which would get called directly from FormBrowse whenever a change would be made to either branches, remotes, or tags, which would then update all 3 Tree objects (Branches, Remotes, and Tags). Right off the bat, this wasn't very efficient, since if only a branch changed, we'd pay the cost of updating all three Trees. Now when I started adding a Submodules Tree, I had a problem because submodule information is retrieved asynchronously by FormBrowse. So I couldn't easily pull this information from the left panel SubmoduleTree. One option was for when FormBrowse gets called back with Submodule info, to have it then push an update to the left panel, somehow getting the new submodule info to it (this info wasn't cached in any way). I discussed this with the team here (starting from here: https://github.com/gitextensions/gitextensions/pull/5617#issuecomment-433020877), and the team suggested I go for a pub/sub solution. Namely, that SubmoduleStatusProvider become a service on which we could subscribe for updates.

So now the idea would be that the SubmoduleTree would subscribe to updates on the SubmoduleStatusProvider; and FormBrowse would do the same to update the Submdoule menu. But for this to work, I would need each Tree in the left panel to be able to update independently of other trees. So I made the changes in https://github.com/gitextensions/gitextensions/commit/c9e47e4b3e79792b1afcc762e0164445c82606a6 to address this. These changes did the following:

* RepoObjectsTree no longer has a ReloadAsync function. Instead, I moved part of this logic into protected Tree.ReloadNodes, a function that would be called by each separate Tree when it needed to reload itself. Previously, RepoObjectsTree.ReloadAsync would call a function "ReloadAsync" on each Tree (Branches, Remotes, and Tags). Now, this same function is passed in as an argument to Tree.ReloadNodes, which it awaits on.
* In order to continue supporting the "push to update" model, I added RepoObjectsTree.RefreshTree, which calls Tree.RefreshTree on each Tree, and these in turn call ReloadNodes(LoadNodesAsync). With this in place, though, we can now move towards having each of these Trees subscribe for changes so they only update when necessary. For example, BranchTree could subscribe for changes to branches, and when a branch change occurs, only this Tree would update, not the other Trees. This is precisely what I do for Submodules.
* The CancellationTokenSource that used to be in RepoObjectsTree also was moved down into Tree, which means that there's one instance of a CancellationToken per Tree, not for all Tree instances. So if a Tree is being updated, but then something cancels it, it will only cancel for that Tree, not the others.
* Clearing Nodes at the start of a Tree.ReloadNodes only clears the node children of that specific Tree. This doesn't affect the other Trees. Furthermore, clearing nodes doesn't cause the flicker you notice. This Nodes hierarchy is not the TreeNode hierarchy that's part of the TreeView control; there is 1:1 mapping, though, and Tree.FillTreeViewNode takes care of efficiently updating the TreeNode hierarchy to match the Node hierarchy, while trying to recycle as many existing TreeNode instances as possible. This hasn't changed at all with my changes.

So basically, my change has made it so that each Tree instance is able to be updated separately from the other Tree instances. This is necessary for Submodules, and eventually will allow us to more efficiently update Branches, Remotes, and Tags in the same way using pub/sub.

The one downside is that, although each Tree instance is now independent in terms of updates, they all share the same TreeView control. So it is true that if all 4 update at the same time, then the same TreeView control gets BeginUpdate/EndUpdate calls on it, which leads to some ugly flicker. However, as we move towards pub/sub for each Tree, we should see less of this type of flickering, since we won't get 4 updates at the same time.

I hope this clarifies the situation more. I'll now respond to your points in light of what I wrote here.

Reply:
> `ReloadNodes` should be an async function. `ThreadHelper.JoinableTaskFactory.RunAsync` should be used only at the beginning of the callstack and then all functions should be async functions if needed. This is how I understand the idea of using JFT.

I think my latest version of ReloadNodes should be okay. It's a protected helper that gets called by each Tree when it needs to reload itself, passing in the async LoadNodesAsync function to get called. I made the changes to ReloadNodes that you pointed out to separate the async call from the UI update so that we no longer ever run the finally block if a CancellationException gets thrown from LoadNodesAsync.

Reply:
> Nodes are cleared by each root node reload action, which causes unpleasant visual effect. I miss ReactJS + JS being single threaded - the code is clearer and easier to write.

As I wrote before, clearing Nodes doesn't cause the flicker. Prior to https://github.com/gitextensions/gitextensions/commit/c9e47e4b3e79792b1afcc762e0164445c82606a6, the root Nodes was getting cleared in RepoObjectsTree.ReloadNodesAsync, so this is the same thing, except we're only clearing the Tree's child Nodes.

Reply:
>I think my latest version of ReloadNodes should be okay. It's a protected helper that gets called by each Tree when it needs to reload itself, passing in the async LoadNodesAsync function to get called. I made the changes to ReloadNodes that you pointed out to separate the async call from the UI update so that we no longer ever run the finally block if a CancellationException gets thrown from LoadNodesAsync.

The reason why it should be async is that the caller is able to await for it. `RefreshTree` is a part of tree's public interface, it invokes indirectly an async call, but does not let the caller to await this async call. Imagine a scenario where you want refresh a tree and then perform another action but only when the refresh action is finished. You can't. You are hiding the async nature of the `RefreshTree` method.

>As I wrote before, clearing Nodes doesn't cause the flicker. Prior to [c9e47e4](https://github.com/gitextensions/gitextensions/commit/c9e47e4b3e79792b1afcc762e0164445c82606a6), the root Nodes was getting cleared in RepoObjectsTree.ReloadNodesAsync, so this is the same thing, except we're only clearing the Tree's child Nodes.

Perhaps there is a room for some kind of throttling.

Reply:
Could return task.whenall(all async functions) for a full reload.

Reply:
Also, could use a semaphore to control begin/end update calls

Reply:
> Also, could use a semaphore to control begin/end update calls or something like it with interlocked.

There is no need for this. BeginUpdate and EndUpdate already work this way internally. You can call BeginUpdate multiple times, and EndUpdate the same number of times. It will only refresh the UI once the last EndUpdate is called. Furthermore, there would be no need for a semaphore since this code must be (and is) called on the main thread.

Reply:
Right.  the goal was to prevent  the need for multiple begin/end update calls.  If you are already in the process of updating and finish another update why call EndUpate.  Why call BeginUpdate.  Only when all update tasks have finished do you call EndUpdate.  

UpdateCount starts at 0
basic sample

void StartUpdate(){
Code here to start update
if(interlocked.Increment(UpdateCount)==1){
tree.beginupdate
}
}

void EndUpdate(){
Code here to End update
if(interlocked.Decrement(UpdateCount)==0){
tree.EndUpdate
}
}

void UpdateBranches(){
StartUpdate()
do stuff here
EndUpdate()
}

void UpdateRemotes(){
StartUpdate()
do stuff here
EndUpdate()
}


void Test{
UpdateBranches()
UpdateRemotes()


If branches is still updating then your UpdateCount would be 2.  After 1 ends it would be 1.  Still no EndUpdate.  Other one finishes.  UpdateCount=0.  EndUpdate is called.  
}

Reply:
> The reason why it should be async is that the caller is able to await for it. `RefreshTree` is a part of tree's public interface, it invokes indirectly an async call, but does not let the caller to await this async call. Imagine a scenario where you want refresh a tree and then perform another action but only when the refresh action is finished. You can't. You are hiding the async nature of the `RefreshTree` method.

Okay, I'll make it async, and see how it goes. Keep in mind, though, that once the other Tree instances (Branches, Remotes, and Tags) go pub/sub, we shouldn't really need RefreshTree anymore.